### PR TITLE
Improve `assert_cairo_test_cases`' feedback

### PR DESCRIPTION
**Pull request by kasperski95**
Improve 'assert_cairo_test_cases' feedback. `CairoCheatcodes` cairo tests can't be run with Protostar in the console. Checking test results with a debugger is annoying.

Before:
Screenshot 2023-01-20 at 16 27 23

After:
Screenshot 2023-01-20 at 16 33 03

### CODE REVIEW

File: tests/integration/conftest.py [hunk 1]
```diff
@@ -81,9 +82,48 @@ def assert_cairo_test_cases(
...
+            if expected_test_case_name not in name_to_test_case_result:
+                lines.append("nothing — no such test case result found")
+                break
+            lines.append(
+                format_test_result(
+                    name_to_test_case_result[expected_test_case_name]
+                ).format_human(LogColorProvider())
+            )
+            lines.append("")
+            break
```
Comment (Arcticae) [inline tests/integration/conftest.py:123]: this whole flow with breaks and continues is confusing. i don't know if it's that important though since it's only used for test output

File: tests/integration/conftest.py [hunk 2]
```diff
@@ -81,9 +82,48 @@ def assert_cairo_test_cases(
...
+    }
+    diff = show_diff_between_cairo_test_cases(
+        name_to_test_case_result, expected, actual
+    )
+    if diff:
+        assert False, diff
     assert actual == expected
 
 
+def show_diff_between_cairo_test_cases(
```
Comment (Arcticae) [inline tests/integration/conftest.py:98]: this doesn't "show" anything. it formats the differences into a string

File: tests/integration/conftest.py [hunk 3]
```diff
@@ -81,9 +82,48 @@ def assert_cairo_test_cases(
...
+        test_result.test_case_name: test_result
+        for test_result in testing_summary.test_results
+        if isinstance(test_result, TestCaseResult)
+    }
+    diff = show_diff_between_cairo_test_cases(
+        name_to_test_case_result, expected, actual
+    )
+    if diff:
+        assert False, diff
     assert actual == expected
```
Comment (Arcticae) [inline tests/integration/conftest.py:95]: isn't this redundant since we're asserting above?

**Review by Arcticae — APPROVED** [review]
A few nitpicks, do as you feel on those, i don't have strong feelings against merging this, since it's only test code.

File: tests/integration/conftest.py [hunk 4]
```diff
@@ -81,9 +82,48 @@ def assert_cairo_test_cases(
...
+    name_to_test_case_result = {
+        test_result.test_case_name: test_result
+        for test_result in testing_summary.test_results
+        if isinstance(test_result, TestCaseResult)
+    }
+    diff = show_diff_between_cairo_test_cases(
+        name_to_test_case_result, expected, actual
+    )
+    if diff:
+        assert False, diff
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply tests/integration/conftest.py:94]: I excluded diff from the condition on purpose, to avoid duplicating diff information.

File: tests/integration/conftest.py (hunk 3 above)
Comment (kasperski95) [reply tests/integration/conftest.py:95]: I am not sure if the diff assertion catches all edge cases. I left the second assertion as a "safety net".

File: tests/integration/conftest.py [hunk 5]
```diff
@@ -81,9 +82,48 @@ def assert_cairo_test_cases(
...
 
 
+def show_diff_between_cairo_test_cases(
+    name_to_test_case_result: dict[str, Any],
+    expected: CairoTestCases,
+    actual: CairoTestCases,
+) -> str:
+    lines: list[str] = []
+    result_types = ["passed", "failed", "broken", "skipped"]
+    for expected_result_type in result_types:
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply tests/integration/conftest.py:105]: I'll change it in the next PR.
